File: src/main/java/com/dansoftware/libraryapp/gui/theme/OsSynchronizedTheme.java
package com.dansoftware.libraryapp.gui.theme;

import com.dansoftware.libraryapp.gui.theme.applier.ThemeApplier;
import com.dansoftware.libraryapp.locale.I18N;
import com.jthemedetecor.OsThemeDetector;
import javafx.application.Platform;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.function.Consumer;

/**
 * A {@link Theme} implementation that synchronizes the App's theme
 * with the System UI Theme.
 *
 * @author Daniel Gyorffy
 */
public class OsSynchronizedTheme extends Theme {

    private static final Logger logger = LoggerFactory.getLogger(OsSynchronizedTheme.class);

    private static final ThemeMeta<OsSynchronizedTheme> THEME_META =
            new ThemeMeta<>(OsSynchronizedTheme.class, () -> I18N.getGeneralValue("theme.sync"), InternalThemeDesigner.INSTANCE);

    static {
        registerTheme(THEME_META);
    }

    private final Consumer<Boolean> osThemeListener;

    private final OsThemeDetector osThemeDetector;
    private volatile Theme currentTheme;
    private final Theme darkTheme;
    private final Theme lightTheme;

    public OsSynchronizedTheme() {
        this.osThemeListener = new SyncFunction(this);
        this.osThemeDetector = OsThemeDetector.getDetector();
        this.osThemeDetector.registerListener(osThemeListener);
        this.darkTheme = new DarkTheme();
        this.lightTheme = new LightTheme();
        this.currentTheme = getCurrentTheme();
    }

    @NotNull
    @Override
    protected ThemeApplier getApplier() {
        return currentTheme.getApplier();
    }

    private Theme getCurrentTheme() {
        return this.osThemeDetector.isDark() ? darkTheme : lightTheme;
    }

    @Override
    protected void onThemeDropped() {
        osThemeDetector.removeListener(osThemeListener);
    }



    private static final class SyncFunction implements Consumer<Boolean> {

        private final OsSynchronizedTheme synchTheme;

        SyncFunction(@NotNull OsSynchronizedTheme synchTheme) {
            this.synchTheme = synchTheme;
        }

        @Override
        public void accept(Boolean isDark) {
            Platform.runLater(() -> {
                if (isDark) {
                    synchTheme.currentTheme = synchTheme.darkTheme;
                    synchTheme.update(synchTheme.lightTheme);
                } else {
                    synchTheme.currentTheme = synchTheme.lightTheme;
                    synchTheme.update(synchTheme.darkTheme);
                }
            });
        }
    }
}

File: boomega-gui/src/main/java/com/dansoftware/boomega/gui/theme/OsSynchronizedTheme.kt
/*
 * Boomega - A modern book explorer & catalog application
 * Copyright (C) 2020-2022  Daniel Gyoerffy
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.dansoftware.boomega.gui.theme

import com.dansoftware.boomega.i18n.api.i18n
import com.jthemedetecor.OsThemeDetector
import javafx.application.Platform
import javafx.scene.Parent
import javafx.scene.Scene
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.lang.ref.WeakReference
import java.util.Collections.synchronizedList
import java.util.function.Consumer

/**
 * Synchronizes the appearance of the UI components based on what's the system's preferred theme (dark/ light).
 */
open class OsSynchronizedTheme : Theme() {

    override val name: String
        get() = i18n("app.ui.theme.sync")

    /**
     * Specifies the [Theme] should be used when the system is in dark mode
     */
    open val darkTheme: Theme
        get() = DarkTheme.INSTANCE

    /**
     * Specifies the [Theme] should be used when the system is in light mode
     */
    open val lightTheme: Theme
        get() = LightTheme.INSTANCE

    /**
     * The list of functions that should be invoked by the os theme detection listener, when
     * the theme changes.
     */
    private val themeChangeEventHandlers =
        synchronizedList(mutableListOf<(isDark: Boolean) -> Boolean>())

    /**
     * The theme-detection listener
     */
    private val themeDetectionListener = Consumer<Boolean> { isDark ->
        Platform.runLater {
            val iterator = themeChangeEventHandlers.iterator()
            while (iterator.hasNext()) {
                val eventHandler = iterator.next()
                val eventHandlerShouldStay = eventHandler(isDark)
                if (!eventHandlerShouldStay) {
                    iterator.remove()
                }
            }
        }
    }

    private inline val osThemeDetector
        get() = OsThemeDetector.getDetector()

    override fun init() {
        super.init()
        osThemeDetector.registerListener(themeDetectionListener)
    }

    override fun destroy() {
        super.destroy()
        logger.debug("OsSynchronizedTheme dropped!")
        logger.debug("Removing os theme listener...")
        osThemeDetector.removeListener(themeDetectionListener)
    }

    override fun apply(scene: Scene) {
        registerScene(WeakReference(scene))
    }

    override fun apply(region: Parent) {
        registerRegion(WeakReference(region))
    }

    override fun deApply(scene: Scene) {
        getAppropriateTheme(osThemeDetector.isDark).deApply(scene)
    }

    override fun deApply(region: Parent) {
        getAppropriateTheme(osThemeDetector.isDark).deApply(region)
    }

    private fun registerScene(scene: WeakReference<Scene>) {
        val applyFunction = fun(isDark: Boolean): Boolean =
            scene.get()?.let {
                getAppropriateTheme(!isDark).deApply(it)
                getAppropriateTheme(isDark).apply(it)
            } != null
        applyFunction(osThemeDetector.isDark)
        themeChangeEventHandlers.add(applyFunction)
    }

    private fun registerRegion(region: WeakReference<Parent>) {
        val applyFunction = fun(isDark: Boolean): Boolean =
            region.get()?.let {
                getAppropriateTheme(!isDark).deApply(it)
                getAppropriateTheme(isDark).apply(it)
            } != null
        applyFunction(osThemeDetector.isDark)
        themeChangeEventHandlers.add(applyFunction)
    }

    private fun getAppropriateTheme(isDark: Boolean) =
        if (isDark) darkTheme else lightTheme

    companion object {

        private val logger: Logger = LoggerFactory.getLogger(OsSynchronizedTheme::class.java)

        /**
         * A global instance of the [OsSynchronizedTheme]
         */
        @JvmField
        val INSTANCE = OsSynchronizedTheme()
    }
}